src/passes/OptimizeInstructions.cpp

```diff
@@ -1995,39 +1995,60 @@ struct OptimizeInstructions
       return;
     }
 
-    // Check whether the cast will definitely fail (or succeed). Look not just
-    // at the fallthrough but all intermediatary fallthrough values as well, as
-    // if any of them has a type that cannot be cast to us, then we will trap,
-    // e.g.
-    //
-    //   (ref.cast $struct-A
-    //     (ref.cast $struct-B
-    //       (ref.cast $array
-    //         (local.get $x)
-    //
-    // The fallthrough is the local.get, but the array cast in the middle
-    // proves a trap must happen.
     Builder builder(*getModule());
-    auto nullType = curr->type.getHeapType().getBottom();
-    {
-      auto** refp = &curr->ref;
-      while (1) {
-        auto* ref = *refp;
 
-        auto result = GCTypeUtils::evaluateCastCheck(ref->type, curr->type);
+    // Look at all the fallthrough values to get the most precise possible type
+    // of the value we are casting. local.tee, br_if, and blocks can all "lose"
+    // type information, so looking at all the fallthrough values can give us a
+    // more precise type than is stored in the IR.
+    Type refType =
+      Properties::getFallthroughType(curr->ref, getPassOptions(), *getModule());
+
+    // If the cast admits null but we know the value is not null, then as a
+    // first step we can tighten up the cast to stop admitting nulls.
```

Comment:
Can we not just apply `cast->type = refType`? That is, the cast must be as refined as the output of `getFallthroughType`, mustn't it? So we can apply not only the nullability but the heap type.

Reply:
We can only do that if the cast would definitely succeed (i.e. when the value's type is a subtype of the cast type), but in that case we prefer to optimize the cast out entirely, not just improve its type. Otherwise we would be changing the semantics of the cast such that it always succeeds, even if that were not originally the case. This could also cause validation failures if we ended up changing the cast type to some unrelated type.

Reply:
I don't understand. Maybe I'm confused about the situation here. Is this transformation not relevant and valid somehow?
```wat
(ref.cast null struct
  (block (result anyref) ;; 'for'get' about the better typing in the children
    (ref.as_non_null
      (ref.cast null $struct.foo ..))))
=>
(ref.cast $struct.foo ;; applied the refined nullability and heap
  (block (result anyref)
    (ref.as_non_null
      (ref.cast null $struct.foo ..))))
```
That seems valid since if we get to the cast then all the constraints above us must have been applied together, so we can use the intersection of them all.

Reply:
Yes, that example's fine, but only because we know the cast will succeed by looking at all the fallthrough values. We already do even better by just replacing the cast entirely in this case.

Here's an example that would go wrong:

```
(ref.cast null $struct
  (block (result anyref)
    (ref.cast i31 ..)))
```

This cast will fail, so if we just replaced it with `ref.cast (ref i31)`, that would be a misoptimization. If this cast is the child of a struct.get, it would also cause a validation failure.

In contrast to arbitrarily changing the heap type here, it is valid to improve the nullability from nullable to nonnullable because that always refines the type and cannot be a misoptimization or cause a validation failure. Furthermore, performing this nullability optimization is not already handled somewhere else.

Reply:
Oh, true, but I think we just need the intersection of it all, including the original cast? The intersection of `(ref i31)` and `(ref null $struct)` is empty, so we could put `(ref.cast none` (empty set) in your example. (And in my example, adding the intersection of the original cast has no effect.)

Reply:
Yes, that would work, but it would be entirely superseded by the optimizations below. In the example you gave, the cast would be removed and the contents of the block would be teed out. In my example, the cast would be turned into a drop followed by `unreachable`.

Reply:
It seems conceptually simpler to me, though, to just take the intersection of all the types (in fallthroughs + the original cast), rather than treat the nullability in a special way + add the code below. Do you disagree?

Reply:
If we had a function in wasm-types.h for taking the meet/glb/intersection of two types, then maybe using that might be simpler, but we don't have a function like that ready to go. How would you feel if I updated the comment to point out that improving the heap type here wouldn't be useful due to the later optimizations?

Reply:
I was thinking that the intersection is what is sort of already computed here - for nullability, improve to non-null, and for heap type, either go to nothing (if unrelated types) or pick the more refined of the two (if related). I see that code in this PR already, I think, so I was hoping there's a good way to refactor it to use that.

But maybe I'm not seeing some practical complication here that prevents it. You're closer to the code. If that seems impractical then sgtm to continue with this PR more or less as is and leave improvements for later?

Reply:
Ok, I think you're right that this would be a nice refactoring in general, even if the refinement of the heap type isn't independently useful in this particular case.

Reply:
PTAL at the last couple of commits ('glb' and 'fewer test changes') for this refactoring.